torch/fx/experimental/symbolic_shapes.py

```diff
@@ -858,13 +858,13 @@ class StatefulSymbolicContext(StatelessSymbolicContext):
     w/r/t different shape_envs, clearing, etc.
     """
     tensor_source: Source = None
-    source_to_symint_node_cache : Dict["TensorPropertySource", SymInt] = None
+    shape_env_to_source_to_symbol_cache : Dict[int, Dict["TensorPropertySource", "sympy.Expr"]] = None
```

Comment:
Question around invariants: when do we have more than one ShapeEnv floating around? Is there a test I can look at that exercises the fact that assuming a single ShapeEnv during this caching is wrong?

It seems a bit weird to me that we have to include a ShapeEnv id in this cache lookup, since:

(1) we should be able to get away with a single ShapeEnv throughout the lifecycle of tracing + compiling a single dynamo graph

(2) I would have thought that this cache (and the tracing context that holds it) would get reset after every dynamo graph break, when start tracing out a fresh graph

Reply:
Check the PR desc :) 

>  Is there a test I can look at that exercises the fact that assuming a single ShapeEnv during this caching is wrong?

`python test/dynamo/test_dynamic_shapes.py -k test_autocast_arguments_binding_dynamic_shapes`

Should do it!

```
So, the fix here is to move the cache up to be a general symbol cache, rather than only a cache for symints.

The initial implementation did that! But then, it ran into some interesting errors when it came to replay. When replaying symbol creation, behaviors would diverge in the new shape env! How could that be? The answer is because creating a new shape_env resulted in us replaying symbol creation... but with a cache from a different shape env! This was short circuiting symbol creation - and so, adding an extra layer to the cache for id(shape_env) fixes the problem.
```

(1) Correct, but we replay for correctness
(2) It does :)

Reply:
When (/ why) do we replay symbol creation? I'll read through the code some more.

> python test/dynamo/test_dynamic_shapes.py -k test_autocast_arguments_binding_dynamic_shapes

So, I tried running this test locally, but "deleting" the per-shape-env cache, just to understand how it would fail. And it still passes for me? I deleted the per-shape-env bits by just changing the `id(self)` key (the id of the ShapeEnv) to 1 everywhere (lol) https://www.internalfb.com/phabricator/paste/view/P900020748